module/zfs/bpobj.c

```diff
@@ -206,155 +206,213 @@ bpobj_is_empty(bpobj_t *bpo)
 	    (!bpo->bpo_havesubobj || bpo->bpo_phys->bpo_num_subobjs == 0));
 }
 
+typedef struct bpobj_info {
+	bpobj_t *bpi_bpo;
+	bpobj_t *bpi_parent;    /* if non NULL then bpi_bpo is a subobj */
+	uint64_t bpi_object;
+	uint64_t bpi_index;
+	list_node_t bpi_node;
+} bpobj_info_t;
+
+static bpobj_info_t *
+bpi_alloc(bpobj_t *bpo, bpobj_t *parent, uint64_t obj, uint64_t index)
+{
+	bpobj_info_t *bpi = kmem_zalloc(sizeof (bpobj_info_t),
+	    KM_SLEEP);
+	bpi->bpi_bpo = bpo;
+	bpi->bpi_parent = parent;
+	bpi->bpi_object = obj;
+	bpi->bpi_index = index;
+	return (bpi);
+}
+
 static int
-bpobj_iterate_impl(bpobj_t *bpo, bpobj_itor_t func, void *arg, dmu_tx_t *tx,
-    boolean_t free)
+bpobj_iterate_impl(bpobj_t *initial_bpo, bpobj_itor_t func, void *arg,
+    dmu_tx_t *tx, boolean_t free)
```

Comment:
This function was a bit long to begin with and now it's become pretty unwieldy.  Let's break it up into two or more functions.  Perhaps a sub-function can deal with the actual freeing of bp's (around lines 252-300).